File: Leetcode/0036-valid-sudoku/solution.py
# ============================================================
# 0036. Valid Sudoku
# https://leetcode.com/problems/valid-sudoku/
# Difficulty: Medium
# Tags: Array, Hash Table, Matrix
# ============================================================


class Solution:
    def isValidSudoku(self, board: list[list[str]]) -> bool:
        """
        Optimal Solution (Array-based Validation)
        Approach: Single pass with boolean arrays for rows, cols, and boxes
        Time:  O(1) — always 81 cells (9x9 fixed board)
        Space: O(1) — three 9x9 boolean arrays (fixed size)
        """
        rows = [[False] * 9 for _ in range(9)]
        cols = [[False] * 9 for _ in range(9)]
        boxes = [[False] * 9 for _ in range(9)]

        for r in range(9):
            for c in range(9):
                if board[r][c] == '.':
                    continue

                d = int(board[r][c]) - 1
                box = (r // 3) * 3 + c // 3

                if rows[r][d] or cols[c][d] or boxes[box][d]:
                    return False

                rows[r][d] = True
                cols[c][d] = True
                boxes[box][d] = True

        return True

    def isValidSudokuHashSet(self, board: list[list[str]]) -> bool:
        """
        Hash Set approach
        Approach: Use sets to track seen digits per row, column, and box
        Time:  O(1) — always 81 cells
        Space: O(1) — 27 sets, each at most 9 elements
        """
        rows = [set() for _ in range(9)]
        cols = [set() for _ in range(9)]
        boxes = [set() for _ in range(9)]

        for r in range(9):
            for c in range(9):
                val = board[r][c]
                if val == '.':
                    continue

                box = (r // 3) * 3 + c // 3

                if val in rows[r] or val in cols[c] or val in boxes[box]:
                    return False

                rows[r].add(val)
                cols[c].add(val)
                boxes[box].add(val)

        return True


# ============================================================
# Test Cases
# ============================================================

if __name__ == "__main__":
    sol = Solution()
    passed = failed = 0

    def test(name: str, got, expected):
        global passed, failed
        if got == expected:
            print(f"\u2705 PASS: {name}")
            passed += 1
        else:
            print(f"\u274c FAIL: {name}")
            print(f"  Got:      {got}")
            print(f"  Expected: {expected}")
            failed += 1

    # Valid board (LeetCode Example 1)
    valid_board = [
        ["5","3",".",".","7",".",".",".","."],
        ["6",".",".","1","9","5",".",".","."],
        [".","9","8",".",".",".",".","6","."],
        ["8",".",".",".","6",".",".",".","3"],
        ["4",".",".","8",".","3",".",".","1"],
        ["7",".",".",".","2",".",".",".","6"],
        [".","6",".",".",".",".","2","8","."],
        [".",".",".","4","1","9",".",".","5"],
        [".",".",".",".","8",".",".","7","9"]
    ]

    # Invalid board (LeetCode Example 2) — duplicate 8 in top-left box
    invalid_board = [
        ["8","3",".",".","7",".",".",".","."],
        ["6",".",".","1","9","5",".",".","."],
        [".","9","8",".",".",".",".","6","."],
        ["8",".",".",".","6",".",".",".","3"],
        ["4",".",".","8",".","3",".",".","1"],
        ["7",".",".",".","2",".",".",".","6"],
        [".","6",".",".",".",".","2","8","."],
        [".",".",".","4","1","9",".",".","5"],
        [".",".",".",".","8",".",".","7","9"]
    ]

    # Board with duplicate in a row
    row_dup_board = [
        ["5","3",".",".","7",".",".",".","5"],  # 5 appears twice in row 0
        ["6",".",".","1","9","5",".",".","."],
        [".","9","8",".",".",".",".","6","."],
        ["8",".",".",".","6",".",".",".","3"],
        ["4",".",".","8",".","3",".",".","1"],
        ["7",".",".",".","2",".",".",".","6"],
        [".","6",".",".",".",".","2","8","."],
        [".",".",".","4","1","9",".",".","5"],
        [".",".",".",".","8",".",".","7","9"]
    ]

    # Board with duplicate in a column
    col_dup_board = [
        ["5","3",".",".","7",".",".",".","."],
        ["6",".",".","1","9","5",".",".","."],
        [".","9","8",".",".",".",".","6","."],
        ["8",".",".",".","6",".",".",".","3"],
        ["4",".",".","8",".","3",".",".","1"],
        ["7",".",".",".","2",".",".",".","6"],
        [".","6",".",".",".",".","2","8","."],
        [".",".",".","4","1","9",".",".","5"],
        ["5",".",".",".","8",".",".","7","9"]  # 5 in col 0 (same as row 0)
    ]

    # Almost empty board
    empty_board = [
        [".",".",".",".",".",".",".",".","."]] * 8 + [
        [".",".",".",".",".",".",".",".","."]
    ]

    # Board with single value
    single_board = [["." for _ in range(9)] for _ in range(9)]
    single_board[4][4] = "5"

    print("=== Array-based Validation (Optimal) ===")

    test("Example 1 (valid)", sol.isValidSudoku(valid_board), True)
    test("Example 2 (invalid box)", sol.isValidSudoku(invalid_board), False)
    test("Row duplicate", sol.isValidSudoku(row_dup_board), False)
    test("Column duplicate", sol.isValidSudoku(col_dup_board), False)
    test("Almost empty board", sol.isValidSudoku(empty_board), True)
    test("Single value board", sol.isValidSudoku(single_board), True)

    print("\n=== Hash Set Approach ===")

    test("HS: Example 1 (valid)", sol.isValidSudokuHashSet(valid_board), True)
    test("HS: Example 2 (invalid box)", sol.isValidSudokuHashSet(invalid_board), False)
    test("HS: Row duplicate", sol.isValidSudokuHashSet(row_dup_board), False)
    test("HS: Column duplicate", sol.isValidSudokuHashSet(col_dup_board), False)
    test("HS: Almost empty board", sol.isValidSudokuHashSet(empty_board), True)
    test("HS: Single value board", sol.isValidSudokuHashSet(single_board), True)

    # Results
    print(f"\n\U0001f4ca Results: {passed} passed, {failed} failed, {passed + failed} total")

File: Leetcode/0036-valid-sudoku/solution.go
package main

import "fmt"

// ============================================================
// 0036. Valid Sudoku
// https://leetcode.com/problems/valid-sudoku/
// Difficulty: Medium
// Tags: Array, Hash Table, Matrix
// ============================================================

// isValidSudoku — Optimal Solution (Array-based Validation)
// Approach: Single pass with boolean arrays for rows, cols, and boxes
// Time:  O(1) — always 81 cells (9x9 fixed board)
// Space: O(1) — three 9x9 boolean arrays (fixed size)
func isValidSudoku(board [][]byte) bool {
	var rows, cols, boxes [9][9]bool

	for r := 0; r < 9; r++ {
		for c := 0; c < 9; c++ {
			if board[r][c] == '.' {
				continue
			}

			d := board[r][c] - '1'
			box := (r/3)*3 + c/3

			if rows[r][d] || cols[c][d] || boxes[box][d] {
				return false
			}

			rows[r][d] = true
			cols[c][d] = true
			boxes[box][d] = true
		}
	}

	return true
}

// isValidSudokuHashSet — Hash Set approach
// Approach: Use maps to track seen digits per row, column, and box
// Time:  O(1) — always 81 cells
// Space: O(1) — 27 maps, each at most 9 entries
func isValidSudokuHashSet(board [][]byte) bool {
	var rows, cols, boxes [9]map[byte]bool

	for i := 0; i < 9; i++ {
		rows[i] = make(map[byte]bool)
		cols[i] = make(map[byte]bool)
		boxes[i] = make(map[byte]bool)
	}

	for r := 0; r < 9; r++ {
		for c := 0; c < 9; c++ {
			val := board[r][c]
			if val == '.' {
				continue
			}

			box := (r/3)*3 + c/3

			if rows[r][val] || cols[c][val] || boxes[box][val] {
				return false
			}

			rows[r][val] = true
			cols[c][val] = true
			boxes[box][val] = true
		}
	}

	return true
}

// ============================================================
// Test Cases
// ============================================================

func main() {
	passed, failed := 0, 0
	test := func(name string, got, expected bool) {
		if got == expected {
			fmt.Printf("\u2705 PASS: %s\n", name)
			passed++
		} else {
			fmt.Printf("\u274c FAIL: %s\n  Got:      %v\n  Expected: %v\n", name, got, expected)
			failed++
		}
	}

	// Helper to create board from strings
	makeBoard := func(rows ...string) [][]byte {
		board := make([][]byte, 9)
		for i, row := range rows {
			board[i] = []byte(row)
		}
		return board
	}

	// Valid board (LeetCode Example 1)
	validBoard := makeBoard(
		"53..7....",
		"6..195...",
		".98....6.",
		"8...6...3",
		"4..8.3..1",
		"7...2...6",
		".6....28.",
		"...419..5",
		"....8..79",
	)

	// Invalid board (LeetCode Example 2) — duplicate 8 in column 0 and top-left box
	invalidBoard := makeBoard(
		"83..7....",
		"6..195...",
		".98....6.",
		"8...6...3",
		"4..8.3..1",
		"7...2...6",
		".6....28.",
		"...419..5",
		"....8..79",
	)

	// Board with duplicate in a row
	rowDupBoard := makeBoard(
		"53..7...5", // 5 appears twice in row 0
		"6..195...",
		".98....6.",
		"8...6...3",
		"4..8.3..1",
		"7...2...6",
		".6....28.",
		"...419..5",
		"....8..79",
	)

	// Board with duplicate in a column
	colDupBoard := makeBoard(
		"53..7....",
		"6..195...",
		".98....6.",
		"8...6...3",
		"4..8.3..1",
		"7...2...6",
		".6....28.",
		"...419..5",
		"5...8..79", // 5 in col 0 (same as row 0)
	)

	// Almost empty board
	emptyBoard := makeBoard(
		".........",
		".........",
		".........",
		".........",
		".........",
		".........",
		".........",
		".........",
		".........",
	)

	// Single value board
	singleBoard := makeBoard(
		".........",
		".........",
		".........",
		".........",
		"....5....",
		".........",
		".........",
		".........",
		".........",
	)

	fmt.Println("=== Array-based Validation (Optimal) ===")

	test("Example 1 (valid)", isValidSudoku(validBoard), true)
	test("Example 2 (invalid box)", isValidSudoku(invalidBoard), false)
	test("Row duplicate", isValidSudoku(rowDupBoard), false)
	test("Column duplicate", isValidSudoku(colDupBoard), false)
	test("Almost empty board", isValidSudoku(emptyBoard), true)
	test("Single value board", isValidSudoku(singleBoard), true)

	fmt.Println("\n=== Hash Set Approach ===")

	// Need fresh boards since byte slices were consumed
	validBoard2 := makeBoard(
		"53..7....",
		"6..195...",
		".98....6.",
		"8...6...3",
		"4..8.3..1",
		"7...2...6",
		".6....28.",
		"...419..5",
		"....8..79",
	)
	invalidBoard2 := makeBoard(
		"83..7....",
		"6..195...",
		".98....6.",
		"8...6...3",
		"4..8.3..1",
		"7...2...6",
		".6....28.",
		"...419..5",
		"....8..79",
	)

	test("HS: Example 1 (valid)", isValidSudokuHashSet(validBoard2), true)
	test("HS: Example 2 (invalid box)", isValidSudokuHashSet(invalidBoard2), false)

	// Results
	fmt.Printf("\n\U0001f4ca Results: %d passed, %d failed, %d total\n", passed, failed, passed+failed)
}
